src/providers/EtherspotContextProvider.tsx

```diff
@@ -540,22 +533,20 @@ const EtherspotContextProvider = ({
     if (onLogout) onLogout();
   }, [setProvider, setProviderAddress, setAccountAddress, onLogout]);
 
-  useEffect(() => {
-    const handleBalanceGet = async () => {
+  const handleGetBalances = useCallback(
+    async (force?: boolean) => {
       if (!sdk || !accountAddress) return;
-      await getSmartWalletBalancesByChain(accountAddress, supportedChains);
-    };
-    handleBalanceGet();
-  }, [supportedChains, sdk, accountAddress]);
 
-  useEffect(() => {
-    const handleKeybasedBalanceGet = async () => {
-      const isSdkConnected = sdk && accountAddress;
-      if (!isSdkConnected || !providerAddress) return;
-      await getKeybasedWalletBalancesPerChain(providerAddress, supportedChains);
-    };
-    handleKeybasedBalanceGet();
-  }, [supportedChains, sdk, providerAddress, accountAddress]);
+      if (!!force || !smartWalletBalanceByChain.length)
+        await getSmartWalletBalancesByChain(accountAddress, supportedChains);
+
+      if (!providerAddress) return;
+
+      if (!!force || !keybasedWalletBalanceByChain.length)
+        await getKeybasedWalletBalancesPerChain(providerAddress, supportedChains);
+    },
+    [sdk, accountAddress, providerAddress, smartWalletBalanceByChain, keybasedWalletBalanceByChain]
+  );
```

Comment:
Hey man! Great work here! The code appears to be well-structured.
I have some small suggestions
1) The function name (handleGetBalances) does not accurately describe what the function does. A better name could be fetchBalances or updateBalances.
2) The function returns undefined. Since it is an async function, it would be better to return a Promise that resolves when the balance have been fetched.

Reply:
Yeah agreed that the name could change. This particular function doesn't need anything to be returned - it's just there to load and store data in the Context, not return data to where it's being called. So it returning `undefined` is intended, since that's the default.